File: organism_kits/swiftui/Sources/HolonsApp/CoaxConfiguration.swift
import Foundation

private let coaxServerEnabledEnv = "OP_COAX_SERVER_ENABLED"
private let coaxServerListenURIEnv = "OP_COAX_SERVER_LISTEN_URI"

public enum CoaxServerTransport: String, Codable, CaseIterable, Identifiable, Sendable {
    case tcp
    case unix

    public var id: String { rawValue }

    public var title: String {
        switch self {
        case .tcp:
            return "TCP"
        case .unix:
            return "Unix socket"
        }
    }

    public var defaultPort: Int {
        switch self {
        case .tcp:
            return 60000
        case .unix:
            return 0
        }
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)

        switch rawValue {
        case Self.tcp.rawValue:
            self = .tcp
        case Self.unix.rawValue:
            self = .unix
        case "webSocket", "restSSE":
            self = .tcp
        default:
            self = .tcp
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

public struct CoaxSettingsSnapshot: Codable, Sendable {
    public var serverTransport: CoaxServerTransport
    public var serverHost: String
    public var serverPortText: String
    public var serverUnixPath: String

    public static let defaultHost = "127.0.0.1"
    public static let defaultUnixPath = defaultCoaxUnixPath()
    public static let defaults = CoaxSettingsDefaults.standard().snapshot

    public init(
        serverTransport: CoaxServerTransport,
        serverHost: String,
        serverPortText: String,
        serverUnixPath: String
    ) {
        self.serverTransport = serverTransport
        self.serverHost = serverHost
        self.serverPortText = serverPortText
        self.serverUnixPath = serverUnixPath
    }
}

public struct CoaxSettingsDefaults: Sendable {
    public let serverHost: String
    public let serverPortText: String
    public let serverUnixPath: String

    public init(
        serverHost: String = CoaxSettingsSnapshot.defaultHost,
        serverPortText: String = String(CoaxServerTransport.tcp.defaultPort),
        serverUnixPath: String
    ) {
        self.serverHost = serverHost
        self.serverPortText = serverPortText
        self.serverUnixPath = serverUnixPath
    }

    public static func standard(
        socketName: String = "organism-holon-coax.sock"
    ) -> CoaxSettingsDefaults {
        CoaxSettingsDefaults(serverUnixPath: defaultCoaxUnixPath(socketName: socketName))
    }

    public var snapshot: CoaxSettingsSnapshot {
        CoaxSettingsSnapshot(
            serverTransport: .tcp,
            serverHost: serverHost,
            serverPortText: serverPortText,
            serverUnixPath: serverUnixPath
        )
    }
}

public struct CoaxLaunchOverrides: Sendable {
    public let isEnabled: Bool?
    public let snapshot: CoaxSettingsSnapshot?

    public init(isEnabled: Bool?, snapshot: CoaxSettingsSnapshot?) {
        self.isEnabled = isEnabled
        self.snapshot = snapshot
    }
}

public func coaxLaunchOverrides(
    environment: [String: String] = ProcessInfo.processInfo.environment,
    defaults: CoaxSettingsDefaults = .standard()
) -> CoaxLaunchOverrides {
    let listenURI = environment[coaxServerListenURIEnv]?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    let enabled = parseBoolOverride(environment[coaxServerEnabledEnv])
    let snapshot = snapshotFromListenURI(listenURI, defaults: defaults)
    return CoaxLaunchOverrides(isEnabled: enabled, snapshot: snapshot)
}

public func resolvedCoaxEnabled(
    storedValue: Bool,
    overrides: CoaxLaunchOverrides
) -> Bool {
    if let enabled = overrides.isEnabled {
        return enabled
    }
    if overrides.snapshot != nil {
        return true
    }
    return storedValue
}

public func parseBoolOverride(_ value: String?) -> Bool? {
    switch value?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
    case "1", "true", "yes", "on":
        return true
    case "0", "false", "no", "off":
        return false
    default:
        return nil
    }
}

public func snapshotFromListenURI(_ listenURI: String) -> CoaxSettingsSnapshot? {
    snapshotFromListenURI(listenURI, defaults: .standard())
}

public func snapshotFromListenURI(
    _ listenURI: String,
    defaults: CoaxSettingsDefaults
) -> CoaxSettingsSnapshot? {
    guard !listenURI.isEmpty else {
        return nil
    }

    if listenURI.hasPrefix("unix://") {
        return CoaxSettingsSnapshot(
            serverTransport: .unix,
            serverHost: defaults.serverHost,
            serverPortText: defaults.serverPortText,
            serverUnixPath: String(listenURI.dropFirst("unix://".count))
        )
    }

    guard listenURI.hasPrefix("tcp://"),
          let components = URLComponents(string: listenURI) else {
        return nil
    }

    let host = components.host?.trimmingCharacters(in: .whitespacesAndNewlines)
    let port = components.port.map(String.init)
    return CoaxSettingsSnapshot(
        serverTransport: .tcp,
        serverHost: (host?.isEmpty == false ? host! : defaults.serverHost),
        serverPortText: port ?? defaults.serverPortText,
        serverUnixPath: defaults.serverUnixPath
    )
}

public func defaultCoaxUnixPath(
    socketName: String = "organism-holon-coax.sock"
) -> String {
    NSTemporaryDirectory() + socketName
}

File: organism_kits/flutter/lib/src/coax_configuration.dart
import 'dart:convert';
import 'dart:io';

enum CoaxSurfaceState {
  off('OFF'),
  saved('SAVED'),
  announced('ANNOUNCED'),
  live('LIVE'),
  error('ERROR');

  const CoaxSurfaceState(this.badgeTitle);

  final String badgeTitle;
}

enum CoaxServerTransport {
  tcp('TCP', 60000),
  unix('Unix socket', 0);

  const CoaxServerTransport(this.title, this.defaultPort);

  final String title;
  final int defaultPort;

  static CoaxServerTransport fromRaw(String value) {
    switch (value.trim().toLowerCase()) {
      case 'unix':
        return CoaxServerTransport.unix;
      case 'tcp':
      case 'websocket':
      case 'restsse':
      default:
        return CoaxServerTransport.tcp;
    }
  }
}

class CoaxSettingsSnapshot {
  const CoaxSettingsSnapshot({
    required this.serverTransport,
    required this.serverHost,
    required this.serverPortText,
    required this.serverUnixPath,
  });

  static const defaultHost = '127.0.0.1';
  static final String defaultUnixPath = defaultCoaxUnixPath();
  static final defaults = CoaxSettingsDefaults.standard().snapshot();

  final CoaxServerTransport serverTransport;
  final String serverHost;
  final String serverPortText;
  final String serverUnixPath;

  Map<String, Object?> toJson() => <String, Object?>{
    'serverTransport': serverTransport.name,
    'serverHost': serverHost,
    'serverPortText': serverPortText,
    'serverUnixPath': serverUnixPath,
  };

  factory CoaxSettingsSnapshot.fromJson(Map<String, dynamic> json) {
    return CoaxSettingsSnapshot(
      serverTransport: CoaxServerTransport.fromRaw(
        (json['serverTransport'] as String?) ?? '',
      ),
      serverHost: (json['serverHost'] as String?) ?? defaultHost,
      serverPortText: (json['serverPortText'] as String?) ?? '60000',
      serverUnixPath: (json['serverUnixPath'] as String?) ?? defaultUnixPath,
    );
  }

  String encode() => jsonEncode(toJson());

  static CoaxSettingsSnapshot decode(String value) {
    if (value.trim().isEmpty) {
      return defaults;
    }
    final decoded = jsonDecode(value);
    if (decoded is! Map<String, dynamic>) {
      return defaults;
    }
    return CoaxSettingsSnapshot.fromJson(decoded);
  }
}

class CoaxSettingsDefaults {
  const CoaxSettingsDefaults({
    this.serverHost = CoaxSettingsSnapshot.defaultHost,
    this.serverPortText = '60000',
    required this.serverUnixPath,
  });

  factory CoaxSettingsDefaults.standard({
    String socketName = 'organism-holon-coax.sock',
  }) {
    return CoaxSettingsDefaults(
      serverUnixPath: defaultCoaxUnixPath(socketName: socketName),
    );
  }

  final String serverHost;
  final String serverPortText;
  final String serverUnixPath;

  CoaxSettingsSnapshot snapshot({
    CoaxServerTransport serverTransport = CoaxServerTransport.tcp,
  }) {
    return CoaxSettingsSnapshot(
      serverTransport: serverTransport,
      serverHost: serverHost,
      serverPortText: serverPortText,
      serverUnixPath: serverUnixPath,
    );
  }
}

class CoaxSurfaceStatus {
  const CoaxSurfaceStatus({
    required this.id,
    required this.title,
    required this.endpoint,
    required this.state,
  });

  final String id;
  final String title;
  final String? endpoint;
  final CoaxSurfaceState state;
}

const int _unixSocketPathMaxBytes = 100;

String coaxTransportTitle(CoaxServerTransport value) => value.title;

int sanitizedPort(String value, int fallback) {
  final parsed = int.tryParse(value.trim());
  if (parsed == null || parsed < 1 || parsed > 65535) {
    return fallback;
  }
  return parsed;
}

String defaultCoaxUnixPath({
  String socketName = 'organism-holon-coax.sock',
}) {
  var current = Directory.systemTemp.absolute.path;
  final seen = <String>{};
  while (seen.add(current)) {
    final candidate = '$current${Platform.pathSeparator}$socketName';
    if (_isWritableDirectory(current) &&
        candidate.codeUnits.length <= _unixSocketPathMaxBytes) {
      return candidate;
    }
    final parent = Directory(current).parent.absolute.path;
    if (parent == current) {
      break;
    }
    current = parent;
  }
  return '${Directory.systemTemp.path}${Platform.pathSeparator}$socketName';
}

bool _isWritableDirectory(String path) {
  final directory = Directory(path);
  if (!directory.existsSync()) {
    return false;
  }

  final probe = File(
    '${directory.path}${Platform.pathSeparator}.coax-write-probe-$pid-${DateTime.now().microsecondsSinceEpoch}',
  );
  try {
    probe.writeAsStringSync('probe');
    probe.deleteSync();
    return true;
  } on FileSystemException {
    return false;
  }
}
